### PR TITLE
(#107) Vehicle Physics

### PR DESCRIPTION
**Pull request by RFSurdsman**
Let's merge the new vehicle physics. We will have to spend lots of time playing around with the values until we get the desired feeling. At the moment it is similar to before with a little bit of drifting.

I've removed all traces of our old vehicle physics. The Car Controller now inherits off the NWH Vehicle Physics

### CODE REVIEW

File: Assets/External/VehiclePhysics/Scripts/Vehicle/VehicleController.cs [hunk 1]
```diff
@@ -116,6 +117,17 @@ public bool Active
...
+        /// is not actually called on the client, so the active status of the car is not changed. This SyncVar hook will ensure the Active
+        /// property setter is called when the SyncVar is updated.
+        /// </summary>
+        /// <param name="active">Whether the car should be active or not.</param>
+        void OnActiveChanged(bool active)
+        {
+            Active = active;
+        }
+
         /// <summary>
```
Comment (RFSurdsman) [inline Assets/External/VehiclePhysics/Scripts/Vehicle/VehicleController.cs]: Read the whole file to understand the purpose of this.

File: ProjectSettings/TimeManager.asset [hunk 2]
```diff
@@ -3,7 +3,7 @@
 --- !u!5 &1
 TimeManager:
   m_ObjectHideFlags: 0
-  Fixed Timestep: 0.02
+  Fixed Timestep: 0.015
```
Comment (RFSurdsman) [inline ProjectSettings/TimeManager.asset]: As recommended by NWH Vehicle Physics.

File: Assets/Infrastructure/CarNetworkTransform.cs [hunk 3]
```diff
@@ -62,30 +58,6 @@ void UpdateRotation(Quaternion realRotation)
...
-        /// <param name="realVerticalInput">The new vertical input.</param>
-        void UpdateVerticalInput(float realVerticalInput)
-        {
-            if (!hasAuthority)
-            {
-                this.realVerticalInput = car.VerticalInput = realVerticalInput;
-            }
-        }
-
         /// <summary>
```
Comment (RFSurdsman) [inline Assets/Infrastructure/CarNetworkTransform.cs]: Input is done differently with NWH compared to before, so the current method of synchronising horizontal and vertical input is impossible (and also impractical). NWH Vehicle Physics has some integration with Photon multiplayer. We can take a look at their code to see how they synchronise the position of players on the network in a separate issue, and perhaps implement something similar to that.

File: Assets/Gameplay/Car/CarPhysicsManager.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        public float SpeedKPH => vehicleController.SpeedKPH;
+
+        /// <summary>
+        /// Cars initially spawn in an inactive state, so they cannot be driven before the race starts.
+        /// Once the race starts, this function is called by the server to allow all players in the race
+        /// to drive their car.
+        /// </summary>
+        /// <param name="active">Whether the car should be active or not.</param>
+        [ClientRpc]
+        public void RpcSetActive(bool active)
```
Comment (Mysterise) [inline Assets/Gameplay/Car/CarPhysicsManager.cs]: `isActive` better name since it's a boolean

File: Assets/Gameplay/Car/CarManager.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+        [SyncVar] GameObject playerGO;
+        public GameObject PlayerGO
+        {
+            get => playerGO;
+            set => playerGO = value;
+        }
+
+        public CarPhysicsManager CarPhysicsManager { get; private set; }
+        public Player Player { get; private set; }
```
Comment (Mysterise) [inline Assets/Gameplay/Car/CarManager.cs]: may want to rename Player to avoid type conflicts

Comment (Mysterise) [reply Assets/Gameplay/Car/CarManager.cs]: suggestion `OwnPlayer`

File: Assets/Infrastructure/Player.cs [hunk 6]
```diff
@@ -15,14 +15,14 @@ public class Player : NetworkBehaviour
         #region Player's Car
 
         [SyncVar] GameObject carGO;
-        CarController car;
-        public CarController Car
+        CarManager car;
```
Comment (Mysterise) [inline Assets/Infrastructure/Player.cs]: maybe should also name this variable `carManager`

File: Assets/Gameplay/Car/CarManager.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        /// <summary>
+        /// Cars initially spawn in an inactive state, so they cannot be driven before the race starts.
+        /// Once the race starts, this function is called by the server to allow all players in the race
+        /// to drive their car.
+        /// </summary>
+        /// <param name="active">Whether the car should be active or not.</param>
+        [ClientRpc]
+        public void RpcSetIsActive(bool active)
```
Comment (Mysterise) [inline Assets/Gameplay/Car/CarManager.cs]: param should also be `isActive`

File: Assets/Infrastructure/Player.cs [hunk 8]
```diff
@@ -15,17 +15,17 @@ public class Player : NetworkBehaviour
         #region Player's Car
 
         [SyncVar] GameObject carGO;
-        CarController car;
-        public CarController Car
+        CarManager carManager;
+        public CarManager CarManager
```
Comment (Mysterise) [inline Assets/Infrastructure/Player.cs]: `OwnCarManager` ?

File: Assets/UX/Camera/BuildingTransparencyRaycaster.cs [hunk 9]
```diff
@@ -13,16 +14,15 @@ public class BuildingTransparencyRaycaster : MonoBehaviour
         /// </summary>
         void Update()
         {
-            CarController car = ClientStateMachine.Singleton.LocalPlayer?.Car;
+            CarManager car = ClientStateMachine.Singleton.LocalPlayer?.CarManager;
 
             if (car != null)
             {
-                foreach (Transform wheelTransform in car.WheelTransforms)
+                foreach (Wheel wheel in car.GetComponent<VehicleController>().Wheels)
```
Comment (Mysterise) [inline Assets/UX/Camera/BuildingTransparencyRaycaster.cs]: expose the car's wheels through the carManager instead?

File: Assets/UX/Car/PlayerBar.cs [hunk 10]
```diff
@@ -12,7 +12,7 @@ public class PlayerBar : MonoBehaviour
     {
         [SerializeField] float playerBarMinDownVelocity = -10; // Minimal velocity needed before applying additional displacement to the bar.
 
-        public CarController Car { get; set; }
+        public CarManager Car { get; set; }
```
Comment (Mysterise) [inline Assets/UX/Car/PlayerBar.cs]: name to `CarManager`

File: Assets/UX/Camera/BuildingTransparencyRaycaster.cs [hunk 11]
```diff
@@ -13,16 +14,15 @@ public class BuildingTransparencyRaycaster : MonoBehaviour
         /// </summary>
         void Update()
         {
-            CarController car = ClientStateMachine.Singleton.LocalPlayer?.Car;
+            CarManager car = ClientStateMachine.Singleton.LocalPlayer?.CarManager;
```
Comment (Mysterise) [inline Assets/UX/Camera/BuildingTransparencyRaycaster.cs]: car to `carManager`

**Pull request merged**